includes/gateways/paypal-standard.php

```diff
@@ -653,3 +653,206 @@ function edd_paypal_link_transaction_id( $transaction_id, $payment_id ) {
+		return;
+	}
+
+	// Localize the refund checkbox label.
+	$label = __( 'Refund Payment in PayPal', 'easy-digital-downloads' );
+
+	?>
+	<script type="text/javascript">
+		jQuery(document).ready(function($) {
+			$('select[name=edd-payment-status]').change(function() {
+				if ( 'refunded' == $(this).val() ) {
+					$(this).parent().parent().append('<input type="checkbox" id="edd-paypal-refund" name="edd-paypal-refund" value="1" style="margin-top:0">');
+					$(this).parent().parent().append('<label for="edd-paypal-refund"><?php echo $label; ?></label>');
+				} else {
+					$('#edd-paypal-refund').remove();
+					$('label[for="edd-paypal-refund"]').remove();
+				}
+			});
+		});
+	</script>
+	<?php
+}
+add_action( 'edd_view_order_details_before', 'edd_paypal_refund_admin_js', 100 );
+
+/**
+ * Possibly refunds a payment made with PayPal Standard or PayPal Express.
+ *
+ * @access public
+ * @since  2.6.0
+ *
+ * @param int $payment_id The current payment ID.
+ * @return void
+ */
+function edd_maybe_refund_paypal_purchase( $payment_id = 0 ) {
+	// Prepare variables.
+	$payment_id = absint( $payment_id );
+	$status  	= isset( $_POST['edd-payment-status'] ) ? stripslashes( strip_tags( $_POST['edd-payment-status'] ) ) : false;
+	$refund     = isset( $_POST['edd-paypal-refund'] ) ? stripslashes( strip_tags( $_POST['edd-paypal-refund'] ) ) : false;
+	$user_id 	= edd_get_payment_user_id( $payment_id );
+	$gateway 	= edd_get_payment_gateway( $payment_id );
+	$processed 	= get_post_meta( $payment_id, '_edd_paypal_refunded', true );
+
+	// If not status is found, return early.
+	if ( ! $status ) {
+		return;
+	}
+
+	// If the status is not set to "refunded", return early.
+	if ( 'refunded' !== $status ) {
+		return;
+	}
+
+	// If the refund box wasn't checked, return false.
+	if ( ! $refund ) {
+		return;
+	}
+
+	// If not PayPal/PayPal Express, return early.
+	if ( 'paypal' !== $gateway && 'paypalexpress' !== $gateway ) {
+		return;
+	}
+
+	// If the payment has already been refunded in the past, return early.
+	if ( $processed ) {
+		return;
+	}
+
+	// Process the refund in PayPal.
+	edd_refund_paypal_purchase( $payment_id );
+}
+add_action( 'edd_updated_edited_purchase', 'edd_maybe_refund_paypal_purchase', 999 );
+
+/**
+ * Refunds a purchase made via PayPal.
+ *
+ * @access public
+ * @since  2.6.0
+ *
+ * @param int $payment_id The current payment ID.
+ * @return void
+ */
+function edd_refund_paypal_purchase( $payment_id = 0 ) {
+	// Verify the payment ID is a valid integer.
+	$payment_id = absint( $payment_id );
+	if ( ! $payment_id ) {
+		echo '<pre>' . var_export( __( 'Payment could not be refunded because the payment ID was invalid.', 'easy-digital-downloads' ), true ) . '</pre>';
+		die;
+	}
+
+	// Grab the transaction ID for the payment. Try notes first since transaction IDs weren't stored in transaction ID field originally.
+	$notes = edd_get_payment_notes( $payment_id );
```

Comment:
Let's retrieve it from the transaction ID helper function first. That function actually looks through notes as well, so probably don't need to do both.